### PR TITLE
binwalk: 2.4.3 -> 3.1.0

### LINKED ISSUE #350300: Bump binwalk to v3
**Project description**

The firmware analysis tool has been rewritten in rust, according to https://github.com/ReFirmLabs/binwalk/issues/691 the first release will be created in the next days.

**Metadata**

* homepage URL: https://github.com/ReFirmLabs/binwalk
* source URL: https://github.com/ReFirmLabs/binwalk
* license: mit
* platforms: unix, linux, darwin, ...

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by felbinger**
Complete re-write in Rust!
- Significantly faster
- Far fewer false positives
- Support for many more file extractors

What's Changed

- Add Dockerfile by @banditopazzo in https://github.com/ReFirmLabs/binwalk/pull/683
- Add support for windows compilation by @ElonGaties in https://github.com/ReFirmLabs/binwalk/pull/689
- main loop refactor: remove redundant queue and no active waiting for results by @banditopazzo in https://github.com/ReFirmLabs/binwalk/pull/690
- Update Dockerfile build dependency install gcc by @kimasplund in https://github.com/ReFirmLabs/binwalk/pull/693
- Edit README.md to be neater and added .gitattributes by @ultimateshadsform in https://github.com/ReFirmLabs/binwalk/pull/696
- dependencies/readme.md: Fix typo by @braillescreen in https://github.com/ReFirmLabs/binwalk/pull/698
- add basic LUKS header support by @nmatt0 in https://github.com/ReFirmLabs/binwalk/pull/699
- Mem cleanup by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/700
- github workflows: first quality check by @banditopazzo in https://github.com/ReFirmLabs/binwalk/pull/701
- Lint fixes by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/705
- add github workflow lint check by @banditopazzo in https://github.com/ReFirmLabs/binwalk/pull/706
- Fixed docker issues by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/707
- Tplink fw rtos sig by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/709
- Added BIN firmware header signature by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/710
- Removed unwrap() from code examples by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/711
- Added support for Autel encoded firmware by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/712
- Ntfs support by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/720
- Apfs support by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/721
- Btrfs support by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/723
- extractors/common: support Chroot path on windows by @135e2 in https://github.com/ReFirmLabs/binwalk/pull/726
- Wince by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/727
- Full search by @devttys0 in https://github.com/ReFirmLabs/binwalk/pull/728

Full Changelog: https://github.com/ReFirmLabs/binwalk/compare/v2.3.4...v3.1.0

closes https://github.com/NixOS/nixpkgs/issues/350300

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by felbinger** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 357991`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>binwalk</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>garble</li>
    <li>ghidra-extensions.ghidraninja-ghidra-scripts</li>
  </ul>
</details>

File: nixos/doc/manual/release-notes/rl-2505.section.md [hunk 1]
```diff
@@ -22,6 +22,9 @@
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
+- `binwalk` was updated to v3.1.0, which has been rewritten in rust. The python module is no longer available.
+  See the release notes of [v3.1.0](https://github.com/ReFirmLabs/binwalk/releases/tag/v3.1.0) for more information.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2505.section.md:26]:
```suggestion
- `binwalk` was updated to 3.1.0, which has been rewritten in rust. The python module is no longer available.
  See the release notes of [3.1.0](https://github.com/ReFirmLabs/binwalk/releases/tag/v3.1.0) for more information.
```

File: pkgs/by-name/bi/binwalk/package.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+  };
+
+  nativeBuildInputs = [ pkg-config ];
+
+  buildInputs = [
+    fontconfig.dev
+    bzip2.out
```
Comment (SuperSandro2000) [inline pkgs/by-name/bi/binwalk/package.nix:29]:
```suggestion
    fontconfig
    bzip2
```
Isn't this choosen automatically?

File: pkgs/by-name/bi/binwalk/package.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "binwalk";
+  version = "3.1.0";
+
+  src = fetchFromGitHub {
+    owner = "ReFirmLabs";
+    repo = "binwalk";
+    rev = "4c71ce48bda3c8f0214d56ad6c18c3a833a8dd29";
```
Comment (SuperSandro2000) [inline pkgs/by-name/bi/binwalk/package.nix:17]:
```suggestion
    rev = "refs/tags/v${version}";
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by felbinger** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 357991`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 6 packages built:</summary>
  <ul>
    <li>binwalk</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>garble</li>
    <li>ghidra-extensions.ghidraninja-ghidra-scripts</li>
  </ul>
</details>

File: pkgs/by-name/bi/binwalk/package.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+rustPlatform.buildRustPackage rec {
+  pname = "binwalk";
+  version = "3.1.0";
+
+  src = fetchFromGitHub {
+    owner = "ReFirmLabs";
+    repo = "binwalk";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-5RgjspfTiiLyIyxr9BEHSlmjlWP02MKuSkMlQS482/E=";
```
Comment (SuperSandro2000) [inline pkgs/by-name/bi/binwalk/package.nix:18]:
```suggestion
    hash = "sha256-em+jOnhCZH5EEJrhXTHmxiwpMcBr5oNU1+5IJ1H/oco=";
```

**Pull request merged**